crates/factor-key-value/src/lib.rs

```diff
@@ -88,17 +89,23 @@ impl Factor for KeyValueFactor {
 
     fn prepare<T: RuntimeFactors>(
         &self,
-        ctx: PrepareContext<T, Self>,
+        mut ctx: PrepareContext<T, Self>,
     ) -> anyhow::Result<InstanceBuilder> {
         let app_state = ctx.app_state();
         let allowed_stores = app_state
             .component_allowed_stores
             .get(ctx.app_component().id())
             .expect("component should be in component_stores")
             .clone();
+        let observe_context = match ctx.instance_builder::<ObserveFactor>() {
+            Ok(factor) => Some(factor.get_observe_context()),
+            Err(spin_factors::Error::NoSuchFactor(_)) => None,
+            Err(e) => return Err(e.into()),
+        };
```

Comment:
Another option to consider here which would reduce the amount of boilerplate in each dependent factor:

```rust
pub struct ObserveContext {
    pub(crate) state: Option<Arc<RwLock<State>>>,
}

impl ObserveContext {
  pub fn from_prepare_context<T: RuntimeFactors, F: Factor>(
    prepare_context: &mut PrepareContext<T, F>
  ) -> ObserveContext {
    let state = // same match expr as above, but getting `Option<Arc<State>>`
    ObserveContext { state }
  }
}
```

Then `reparent_tracing_span` could handle `self.state == None` internally.